pkg/agent/core/metrics.go

```diff
@@ -4,78 +4,143 @@ package core
-// output from node_exporter/vector is so small anyways.
-func ReadMetrics(nodeExporterOutput []byte, loadPrefix string) (m Metrics, err error) {
-	lines := strings.Split(string(nodeExporterOutput), "\n")
-
-	getField := func(linePrefix, dontMatch string) (float32, error) {
-		var line string
-		for _, l := range lines {
-			if strings.HasPrefix(l, linePrefix) && (len(dontMatch) == 0 || !strings.HasPrefix(l, dontMatch)) {
-				line = l
-				break
-			}
-		}
-		if line == "" {
-			return 0, fmt.Errorf("No line in metrics output starting with %q", linePrefix)
-		}
+// It is parameterized by the config 'C' passed to the parsing function, to allow things like metric
+// names to be defined externally, rather than compile-time constants.
+type FromPrometheus interface {
+	fromPrometheus(map[string]*promtypes.MetricFamily) error
+}
 
-		fields := strings.Fields(line)
-		if len(fields) < 2 {
-			return 0, fmt.Errorf(
-				"Expected >= 2 fields in metrics output for %q. Got %v",
-				linePrefix, len(fields),
-			)
-		}
+// ParseMetrics reads the prometheus text-format content, parses it, and uses M's implementation of
+// FromPrometheus to populate it before returning.
+func ParseMetrics(content io.Reader, metrics FromPrometheus) error {
+	var parser promfmt.TextParser
+	mfs, err := parser.TextToMetricFamilies(content)
+	if err != nil {
+		return fmt.Errorf("failed to parse content as prometheus text format: %w", err)
+	}
+
+	if err := metrics.fromPrometheus(mfs); err != nil {
+		return fmt.Errorf("failed to extract metrics: %w", err)
+	}
+
+	return nil
+}
+
+func extractFloatGauge(mf *promtypes.MetricFamily) (float64, error) {
+	if mf.GetType() != promtypes.MetricType_GAUGE {
+		return 0, fmt.Errorf("wrong metric type: expected %s but got %s", promtypes.MetricType_GAUGE, mf.GetType())
+	} else if len(mf.Metric) != 1 {
+		return 0, fmt.Errorf("expected 1 metric, found %d", len(mf.Metric))
+	}
+
+	return mf.Metric[0].GetGauge().GetValue(), nil
+}
 
-		v, err := strconv.ParseFloat(fields[1], 32)
-		if err != nil {
-			return 0, fmt.Errorf(
-				"Error parsing %q as float for line starting with %q: %w",
-				fields[1], linePrefix, err,
-			)
+// Helper function to return an error for a missing metric
+func missingMetric(name string) error {
+	return fmt.Errorf("missing expected metric %s", name)
+}
+
+// fromPrometheus implements FromPrometheus, so SystemMetrics can be used with ParseMetrics.
+func (m *SystemMetrics) fromPrometheus(mfs map[string]*promtypes.MetricFamily) error {
+	ec := &erc.Collector{}
+
+	getFloat := func(metricName string) float64 {
+		if mf := mfs[metricName]; mf != nil {
+			f, err := extractFloatGauge(mf)
+			ec.Add(err) // does nothing if err == nil
+			return f
+		} else {
+			ec.Add(missingMetric(metricName))
+			return 0
 		}
-		return float32(v), nil
 	}
 
-	m.LoadAverage1Min, err = getField(loadPrefix+"load1", loadPrefix+"load15")
-	if err != nil {
-		return
+	names := struct {
```

Comment:
I'd just have 4 constants, but up to you.

Reply:
Reworked this so we're the strings are directly passed to `getFloat`, similar to how LFC metrics parsing is now implemented, see 66b92893a804fc0f17b84f7c67358d9610573cb1.